### PR TITLE
Fix/120683 116980 attendee data and woo bulk complete

### PR DESCRIPTION
**Pull request by cliffordp**
https://central.tri.be/issues/120683 and https://central.tri.be/issues/116980

### CODE REVIEW

File: readme.txt [hunk 1]
```diff
@@ -37,7 +37,10 @@ We're always interested in your feedback and our [premium forums](https://theeve
 
 = [1.2.1] 2019-TBD =
 
+* Fix - Attendee Information from Event Tickets Plus now appears correctly at time of emailing [120683]
+* Fix - Bulk completing WooCommerce Orders only includes each Order's applicable PDF attachments [116980]
 * Fix - No longer throws fatal error when Permalinks are disabled
+* Tweak - Now requires Event Tickets Plus version 4.7 or newer.
```
Comment (sc0ttkclark) [inline readme.txt]:
```suggestion
```

File: readme.txt [hunk 2]
```diff
@@ -37,7 +37,10 @@ We're always interested in your feedback and our [premium forums](https://theeve
 
 = [1.2.1] 2019-TBD =
 
+* Fix - Attendee Information from Event Tickets Plus now appears correctly at time of emailing [120683]
```
Comment (sc0ttkclark) [inline readme.txt]:
```suggestion
* Tweak - Now requires Event Tickets Plus version 4.7 or newer.
* Fix - Attendee Information from Event Tickets Plus now appears correctly at time of emailing [120683]
```

File: tribe-ext-pdf-tickets.php [hunk 3]
```diff
@@ -1441,22 +1503,27 @@ protected function output_pdf( $html, $file_name, $dest = 'F' ) {
...
-			 * Empty the output buffer to ensure the website page's HTML is not
-			 * included by accident.
+			 * Empty the output buffer to ensure the website page's HTML is not included by accident.
 			 *
 			 * @link https://mpdf.github.io/what-else-can-i-do/capture-html-output.html
 			 * @link https://stackoverflow.com/a/35574170/893907
 			 */
 			@ob_clean();
 
+			/** @var Mpdf $mpdf */
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]: You don't need this if the get_mpdf has it's own return phpdoc

```suggestion
```

File: tribe-ext-pdf-tickets.php [hunk 4]
```diff
@@ -50,12 +52,19 @@ class Tribe__Extension__PDF_Tickets extends Tribe__Extension {
 		public $pdf_retry_url_query_arg_key = 'tribe_ext_pdf_tickets_retry';
 
 		/**
-		 * An array of the absolute file paths of the PDF(s) to be attached
-		 * to the ticket email.
+		 * A running count of the number of PDFs, such as if multiple tickets were purchased within a single Order.
+		 *
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]:
```suggestion
		 *
		 * @since 1.2.1
		 *
```

File: tribe-ext-pdf-tickets.php [hunk 5]
```diff
@@ -223,18 +240,18 @@ public function init() {
...
+			add_action( 'event_tickets_rsvp_ticket_created', [ $this, 'do_upload_pdf' ], 50, 1 );
 
 			// Event Tickets: Tribe PayPal
 			add_action( 'event_tickets_tpp_attendee_created', [ $this, 'do_upload_pdf' ], 50, 1 );
 
+			// Event Tickets Plus: Easy Digital Downloads (EDD)
+			add_action( 'eddtickets-send-tickets-email', [ $this, 'do_upload_pdf' ], 50, 1 );
+
 			// Event Tickets Plus: WooCommerce
-			add_action( 'event_ticket_woo_attendee_created', [ $this, 'do_upload_pdf' ], 50, 1 );
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]: What was the reason for removing this? Just making sure it was intentional

File: tribe-ext-pdf-tickets.php [hunk 6]
```diff
@@ -261,10 +278,13 @@ public function create_pdf_file_creation_deletion_triggers() {
...
 		 * Order ID, not the Attendee ID.
 		 *
-		 * @param int|string $order_id
+		 * @since 1.1.0
 		 *
 		 * @see Tribe__Tickets_Plus__Commerce__WooCommerce__Main::get_attendees_by_id()
 		 * @see Tribe__Tickets_Plus__Commerce__WooCommerce__Main::send_tickets_email()
+		 *
+		 * @param int|string $order_id
+		 *
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]:
```suggestion
```

File: tribe-ext-pdf-tickets.php [hunk 7]
```diff
@@ -827,6 +857,43 @@ public function do_upload_pdf( $attendee_id, $email = true ) {
...
 		}
 
+		/**
+		 * Given a WooCommerce Ticket's Attendee ID, get the WooCommerce Order ID and add the PDF file to the
+		 * attachments array with the Order ID as the array key.
+		 *
+		 * If the attachments array has non-matching array key(s), clear out attachments array and add this new PDF.
+		 * We do this to avoid the issue of bulk Complete orders in WooCommerce wp-admin causing email 1 to have its PDF
+		 * and then email 2 including both email 2's PDF plus email 1's PDF (yikes!).
+		 *
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]:
```suggestion
		 *
		 * @since 1.2.1
		 *
```

File: tribe-ext-pdf-tickets.php [hunk 8]
```diff
@@ -827,6 +857,43 @@ public function do_upload_pdf( $attendee_id, $email = true ) {
...
+		 * @param $attendee_id
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]:
```suggestion
		 * @param int $attendee_id Attendee ID.
```

File: tribe-ext-pdf-tickets.php [hunk 9]
```diff
@@ -827,6 +857,43 @@ public function do_upload_pdf( $attendee_id, $email = true ) {
...
+		 * @param $file_name
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]:
```suggestion
		 * @param string $file_name PDF file name.
```

File: tribe-ext-pdf-tickets.php [hunk 10]
```diff
@@ -827,6 +857,43 @@ public function do_upload_pdf( $attendee_id, $email = true ) {
...
+			foreach ( $this->attachments_array as $key => $value ) {
+				if ( 0 !== strpos( $key, 'woo_' . $woo_order_id ) ) {
+					$this->attachments_array = [];
+					$this->attachment_count  = 1;
+					break;
+				}
+			}
+
+			// add to array with key like `woo_84202_1`
+			$this->attachments_array[$order_id_array_key . '_' . $this->attachment_count] = $file_name;
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]:
```suggestion
			$this->attachments_array[ $order_id_array_key . '_' . $this->attachment_count ] = $file_name;
```

File: tribe-ext-pdf-tickets.php [hunk 11]
```diff
@@ -1254,10 +1322,7 @@ public function ticket_link( $attendee_id ) {
 				return '';
 			}
 
-			$output = sprintf(
-				'<a href="%s"',
-				$url
-			);
+			$output = sprintf( '<a href="%s"', $url);
```
Comment (sc0ttkclark) [inline tribe-ext-pdf-tickets.php]: This should have esc_url() in it and you should remove the esc_url() being called above.

File: tribe-ext-pdf-tickets.php (hunk 5 above)
Comment (cliffordp) [reply tribe-ext-pdf-tickets.php]: I added code comments to explain why different than other ticket providers

File: tribe-ext-pdf-tickets.php (hunk 11 above)
Comment (cliffordp) [reply tribe-ext-pdf-tickets.php]: `esc_url()` happened upstream to ensure the text is actually a URL - and if not then don't output the HTML for a link

Comment (sc0ttkclark) [reply tribe-ext-pdf-tickets.php]: You can set the `$pdf_link = $this->get_pdf_link( $attendee_id );` then check `if ( empty( esc_url_raw( $url ) )` and then output this as 
```suggestion
			$output = sprintf( '<a href="%s"', esc_url( $url ) );
```

**Comment by sc0ttkclark** [comment]
Merge after QA passes

**Comment by cliffordp** [comment]
@sc0ttkclark fresh code review after https://central.tri.be/issues/120683 failed QA - now passes self-QA

**Pull request merged**